### PR TITLE
tests: input: add a gpio-kbd-matrix test suite

### PR DESCRIPTION
**Pull request by fabiobaltieri**
Test [coverage](https://github.com/zephyrproject-rtos/zephyr/assets/891546/1f2ed5e3-75b8-44c1-9e70-b9abf5fbc9b3).

---

Add an initial test suite for the gpio-kbd-matrix driver.

### CODE REVIEW

**Review by keith-zephyr — COMMENTED** [review]
Can you

File: tests/drivers/input/gpio_kbd_matrix/src/main.c [hunk 1]
```diff
@@ -0,0 +1,398 @@
...
+	int val;
+	int event_count;
+} test_event_data;
+
+static int last_checked_event_count;
+
+static void assert_no_new_events(void)
+{
+	zassert_equal(last_checked_event_count, test_event_data.event_count);
+}
```
[the remark this answers is not in the document]
Comment (fabiobaltieri) [reply tests/drivers/input/gpio_kbd_matrix/src/main.c:170]: hey that's a great idea, I'll fix it in https://github.com/zephyrproject-rtos/zephyr/pull/65168 as well, that bit is copy pasted, should probably look into moving these two a common code file but I'd rather do it in a followup at this stage

File: tests/drivers/input/gpio_kbd_matrix/src/main.c [hunk 2]
```diff
@@ -0,0 +1,398 @@
...
+#define COL_COUNT 2
+
+static uint8_t test_rows[KBD_DEV_COUNT][COL_COUNT];
+static int scan_set_count[KBD_DEV_COUNT];
+
+static void gpio_kbd_scan_set_row(const struct device *dev, uint8_t row)
+{
+	if (dev == dev_interrupt) {
+		gpio_emul_input_set(dev_gpio, 0, !(row & BIT(0)));
+		gpio_emul_input_set(dev_gpio, 1, !(row & BIT(1)));
```
[the remark this answers is not in the document]
Comment (fabiobaltieri) [reply tests/drivers/input/gpio_kbd_matrix/src/main.c:40]: yeah that's a good idea, I guess I got a bit lazy on that one, added a bunch of defines for these :-)

File: tests/drivers/input/gpio_kbd_matrix/src/main.c [hunk 3]
```diff
@@ -0,0 +1,398 @@
...
+	} else if (dev == dev_scan) {
+		gpio_emul_input_set(dev_gpio, 9, !(row & BIT(0)));
+		gpio_emul_input_set(dev_gpio, 8, !(row & BIT(1)));
+		return;
+	}
+
+	TC_PRINT("unknown device: %s\n", dev->name);
+}
+
+void input_kbd_matrix_drive_column_hook(const struct device *dev, int col)
```
[the remark this answers is not in the document]
Comment (fabiobaltieri) [reply tests/drivers/input/gpio_kbd_matrix/src/main.c:55]: whish I could say it was planned, but I just got lucky

**Review by gmarull — DISMISSED** [review]

**Comment by fabiobaltieri** [comment]
Ported over a couple comments from https://github.com/zephyrproject-rtos/zephyr/pull/66174

**Review by keith-zephyr — APPROVED** [review]

**Review by gmarull — APPROVED** [review]

**Pull request merged**